File: 01-Programming-Languages/R/Data-Structures/Stack/valid_parentheses.R
# Problem: Valid Parentheses
# Description: Check if a string of brackets is properly closed and nested.
#
# Approach:
# - Use a stack to track opening brackets
# - Match closing brackets with the top of the stack
#
# Time Complexity: O(n)
# Space Complexity: O(n)
#
# Example:
# Input: "()[]{}"
# Output: TRUE
# Input: "(]"
# Output: FALSE

is_valid <- function(s) {
  stack <- c()
  pairs <- c("(" = ")", "{" = "}", "[" = "]")
  for (i in 1:nchar(s)) {
    ch <- substr(s, i, i)
    if (ch %in% names(pairs)) {
      stack <- c(stack, ch)
    } else {
      if (length(stack) == 0) return(FALSE)
      top <- stack[length(stack)]
      stack <- stack[-length(stack)]
      if (pairs[top] != ch) return(FALSE)
    }
  }
  length(stack) == 0
}

test_cases <- c("()[]{}", "(]", "([)]", "{[]}", "")
for (s in test_cases) {
  cat(sprintf('"%s" -> %s\n', s, ifelse(is_valid(s), "TRUE", "FALSE")))
}

File: 01-Programming-Languages/Scala/Data-Structures/Stack/ValidParentheses.scala
// Problem: Valid Parentheses
// Description: Check if a string of brackets is properly closed and nested.
//
// Approach:
// - Use a stack to track opening brackets
// - Match closing brackets with the top of the stack
//
// Time Complexity: O(n)
// Space Complexity: O(n)
//
// Example:
// Input: "()[]{}"
// Output: true
// Input: "(]"
// Output: false

object ValidParentheses {
  def isValid(s: String): Boolean = {
    val stack = scala.collection.mutable.Stack[Char]()
    val pairs = Map('(' -> ')', '{' -> '}', '[' -> ']')
    for (ch <- s) {
      if (pairs.contains(ch)) {
        stack.push(ch)
      } else {
        if (stack.isEmpty) return false
        val top = stack.pop()
        if (pairs(top) != ch) return false
      }
    }
    stack.isEmpty
  }

  def main(args: Array[String]): Unit = {
    val testCases = Seq("()[]{}", "(]", "([)]", "{[]}", "")
    for (s <- testCases) {
      println(s""""$s" -> ${isValid(s)}""")
    }
  }
}
